### ISSUE TITLE
Multi-Node Microk8s

### ISSUE THREAD

**Issue by vinhdiesal** [ubuntu/microk8s#1255]
I have a few computers I want to create a Kubernetes clusters. The problem with kubeadm is that it blows up my computer (i have a Razer) running on Ubuntu 20.04 and it is not compatible with the Mac OSx machines. 
Apparently minikube doesn't support multi-node clustering, so the closest solution I have in making this Kubernetes cluster work is with MicroK8s. 
I'm using the Multiclass with Ubuntu 18.04 on all of my machines and looking for instructions on how to join these different computers on the same wireless router. 
Do I need an active SSH connection? Do I use the VM IP address? 

Thanks,

**Comment by ktsakalozos** [comment]
Hi @vinhdiesal, you should create a network across all nodes so the VMs running into them could see each other without a NAT.

**Comment by vinhdiesal** [comment]
Hi, Currently all my machines have the appropriate replicated SSH -keys to connect to each other without a password. 
What I tried doing was adding nodes to the microk8s kubernetes cluster. 
Here's my procedure: 
On one machine I typed in microk8s add-node which gave me a join script which included a key and ipaddress. 
I performed an ssh into my computer and logged in the multipass shell with microk8s and typed in 
`microk8s join ipaddress XXXXXXXXXXXXXXXXXXXXXXXXXXX`
I got the following errors: 

```Traceback (most recent call last):
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/requests/adapters.py", line 376, in send
    timeout=timeout
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 610, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 283, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
requests.packages.urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='192.168.1.191', port=25000): Max retries exceeded with url: /cluster/api/v1.0/join (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f54ed87e198>: Failed to establish a new connection: [Errno 113] No route to host',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/snap/microk8s/1378/scripts/cluster/join.py", line 623, in <module>
    join_etcd(connection_parts)
  File "/snap/microk8s/1378/scripts/cluster/join.py", line 581, in join_etcd
    info = get_connection_info(master_ip, master_port, token, callback_token=callback_token)
  File "/snap/microk8s/1378/scripts/cluster/join.py", line 74, in get_connection_info
    verify=False)
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/requests/api.py", line 107, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/requests/api.py", line 53, in request
    return session.request(method=method, url=url, **kwargs)
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/requests/sessions.py", line 480, in request
    resp = self.send(prep, **send_kwargs)
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/requests/sessions.py", line 588, in send
    r = adapter.send(request, **kwargs)
  File "/snap/microk8s/1378/usr/lib/python3/dist-packages/requests/adapters.py", line 437, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='192.168.1.191', port=25000): Max retries exceeded with url: /cluster/api/v1.0/join (Caused by NewConnectionError('<requests.packages.urllib3.connection.VerifiedHTTPSConnection object at 0x7f54ed87e198>: Failed to establish a new connection: [Errno 113] No route to host',))```

**Comment by jdpc86** [comment]
...hmm I happen to run into the same issue today (master mk8s running locally, multipass vm mk8s)...
join via the multipass vm bridge network was hitting the "No route to host", I tried to telnet ip:25000 from vm and it did not work...

it turned out to be a firewall issue, opened it up on host (opened for all ports)...and after that microk8s join worked and a cluster with 2 nodes created...

Although I'm running into a new issue when scaling up replicas, it's falling back to the master instead of distribute over 2 nodes due to "MissingClusterDNS  50m (x20 over 69m)  kubelet, jd        pod: "kubernetes-bootcamp-6f6656d949-ff9vv_default(3ac1def5-62af-4065-8dce-b2b63b3e7605)". kubelet does not have ClusterDNS IP configured and cannot create Pod using "ClusterFirst" policy. Falling back to "Default" policy."

from what I could see, this was due to kubelet did not have --cluster-dns option set to the kube-dns cluster ip? I'm looking to find where i could have that set up to pass it into kubelet command...

**Comment by vinhdiesal** [comment]
I have opened the ports on the firewall on the host machine via `ufw` (`any to address port, any to port, etc`)however, still no success on getting the microk8s to communicate with each other. When I tested out the `ssh username@hostname` it establishes connection, however when I test out `ssh username@hostname -p port number` it refuses to establish connections. 
Can you help me get the machines communicate with each other through the different ports, I think that alone may help `solve` alot of my problems?

**Comment by vinhdiesal** [comment]
I solved the problem by re-installing Ubuntu 20.04 and microk8s and I was able to get the multi-node cluster up and running. On the `master` I performed a `microk8s add-node` then I performed an ssh to my other machines and pasted the `microk8s join <>:2500/<token>`
Thanks!
